gobblin-service/src/main/java/org/apache/gobblin/service/modules/flow/BaseFlowToJobSpecCompiler.java

```diff
@@ -0,0 +1,274 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.service.modules.flow;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nonnull;
+
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.Timer;
+import com.google.common.base.Optional;
+import com.google.common.collect.Maps;
+import com.typesafe.config.Config;
+import com.typesafe.config.ConfigValueFactory;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.gobblin.configuration.State;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.apache.gobblin.metrics.MetricContext;
+import org.apache.gobblin.metrics.Tag;
+import org.apache.gobblin.runtime.api.Spec;
+import org.apache.gobblin.runtime.api.SpecCompiler;
+import org.apache.gobblin.runtime.api.TopologySpec;
+import org.apache.gobblin.configuration.ConfigurationKeys;
+import org.apache.gobblin.instrumented.Instrumented;
+import org.apache.gobblin.runtime.job_catalog.FSJobCatalog;
+import org.apache.gobblin.service.ServiceConfigKeys;
+import org.apache.gobblin.service.ServiceMetricNames;
+import org.apache.gobblin.util.ConfigUtils;
+import org.apache.gobblin.annotation.Alpha;
+import org.apache.gobblin.runtime.api.FlowSpec;
+import org.apache.gobblin.runtime.api.JobSpec;
+import org.apache.gobblin.runtime.api.JobTemplate;
+import org.apache.gobblin.runtime.api.SpecExecutor;
+import org.apache.gobblin.runtime.api.SpecProducer;
+import org.apache.gobblin.runtime.api.SpecNotFoundException;
+import org.apache.gobblin.runtime.job_spec.ResolvedJobSpec;
+
+import lombok.Getter;
+import lombok.Setter;
+
+// Provide base implementation for constructing multi-hops route.
+@Alpha
+public abstract class BaseFlowToJobSpecCompiler implements SpecCompiler{
+
+  // The following two maps are required to be update in atomic, ideally.
+  // Since {@link SpecCompiler} is an {@link SpecCatalogListener}, it is expected that anyS Spec change should be reflected
+  // to these data structures.
+  @Getter
+  @Setter
+  protected final Map<URI, TopologySpec> topologySpecMap;
+  /**
+   * Mapping between {@link SpecExecutor}'s URI and
+   * the corresponding {@link SpecExecutor}'s reference to its {@link SpecProducer}, which would be used to construct
+   * output of FlowCompiler.
+   * This map is constructed mainly for querying purpose.
+   */
+  @Getter
+  @Setter
+  protected final Map<URI, SpecExecutor> specExecutorMap;
+
+
+  /**
+   * Mapping between each FlowEdge and a list of applicable Templates.
+   * Compiler should obtain this Map info from {@link org.apache.gobblin.service.modules.orchestration.Orchestrator} or even higher level,
```

Comment:
Orchestrator and Compiler should not exchange this information. They are relatively independent entities.